Makefile

```diff
@@ -6,7 +6,7 @@ test:
 test-all:
 	CGO_ENABLED=1 go test -race ./...
 benchmark:
-	CGO_ENABLED=0 go test -run='^$$' -bench=. -benchtime=100ms ./...
+	CGO_ENABLED=0 go test -count=10 -run='^$$' -bench=. -benchtime=1ns ./... | tee benchmark.txt
```

Comment:
i don't think `-benchtime=1ns` results in a realistic run that runs in realistic results.  what's wrong with the default?

Reply:
You're right and I set it to 1ms instead. The idea is to not take hours to run. With 1ms it takes in the order of 10-15 minutes.

Reply:
(default is 1 second)

Reply:
though I don't remember which ones, I'm pretty sure we have benchmarks that don't produce reliable output with such an aggressive benchtime.

Reply:
The ones that seem to vary the most are:
pkg:github.com/grafana/metrictank/expr/tagquery & pkg:github.com/grafana/metrictank/idx 
~20% faster in average when running for 100ms instead of 1ms

pkg:github.com/grafana/metrictank/mdata/cache
x3 to x90 slower in average when running for 100ms instead of 1ms

Using `-benchtime=1ms` it takes 15 minutes for the entire testsuite to run here.
Using `-benchtime=100ms` it takes 40 minutes for the entire testsuite to run here.

Note that the fact that the average timings are different in both scenarii does not necessarily matter:
1) we are more interested in the variation of the results over time/commits rather than the absolute numbers for a given bench
2) the better `-benchtime` depends on the context the functions we benchmark are used: if we use a function very repeatedly in production then a higher benchtime is slightly more representative of a real world workload; if the function is used infrequently or its repetitions are very much spaced in time then a smaller benchtime might be more appropriate (so that we don't inadvertendly take advantage of the caches)

Reply:
There seems to be around 639 benchmarks run in total:

> go test -list "Benchmark*" ./... | grep Bench | wc -l
> 639

Reply:
i seem to recall some of our benchmarks needing some time to settle before the workload we expect to measure starts really being executed in a significant way. but the only way to surface those would be to study all of them.
we could get rid of all the auto generated messagepack benchmarks.

Reply:
> the better -benchtime depends on the context the functions we benchmark are used: if we use a function very repeatedly in production then a higher benchtime is slightly more representative of a real world workload; if the function is used infrequently or its repetitions are very much spaced in time then a smaller benchtime might be more appropriate (so that we don't inadvertendly take advantage of the caches)

interesting. seems to make sense. so by that logic (and also to address my concern) we may want to use different benchtimes for different benchmarks.

Reply:
> > the better -benchtime depends on the context the functions we benchmark are used: if we use a function very repeatedly in production then a higher benchtime is slightly more representative of a real world workload; if the function is used infrequently or its repetitions are very much spaced in time then a smaller benchtime might be more appropriate (so that we don't inadvertendly take advantage of the caches)
> 
> interesting. seems to make sense. so by that logic (and also to address my concern) we may want to use different benchtimes for different benchmarks.

that's what I feel yes, however I did not find a way to do that.

Reply:
Extra datapoint: running with -benchtime=1s makes the test suite take a little over 2 hours.

Reply:
> `pkg:github.com/grafana/metrictank/mdata/cache`
> x3 to x90 slower in average when running for 100ms instead of 1ms
> 

I see why this specific one is broken: the benchmarks modify `b.N` with code such as:
`b.N = b.N - (b.N % 4)`

Reply:
> `pkg:github.com/grafana/metrictank/mdata/cache`
> x3 to x90 slower in average when running for 100ms instead of 1ms
> 

I see why this specific one is broken: the benchmarks use `b.N` for the number of chunks with code such as:
`chunks := generateChunks(b, 10, uint32(b.N), 10)`
What is tested depends on `-benchtime` because of that.

Reply:
every benchmark should execute the main code being tested b.N times. The code you refer to does exactly that. Not sure why that is a problem?

Reply:
> every benchmark should execute the main code being tested b.N times. The code you refer to does exactly that (by creating b.N chunks and adding them all). Not sure why that is a problem?

The main code being tested is changing at every iteration (for every b.N) because the number of chunks changes because b.N changes. So we are testing 1 time b.N different pieces of code instead of b.N times 1 piece of code.

Reply:
> Using `-benchtime=1ms` it takes 15 minutes for the entire testsuite to run here.
> Using `-benchtime=100ms` it takes 40 minutes for the entire testsuite to run here.

Using `-benchtime=1000ms` it takes 2h30 for the entire testsuite to run here.

Reply:
Benchmarks for 1ms, 100ms, 1000ms are available at:
https://github.com/grafana/metrictank/pull/1496#issuecomment-543684294

Reply:
decision is to use  `-benchtime=100ms`